node/src/components/linear_chain_sync.rs

```diff
@@ -400,33 +419,71 @@ where
     ) -> Effects<Self::Event> {
         match event {
             Event::Start(init_peer) => {
-                match self.state {
+                match &self.state {
                     State::None => {
                         // No syncing configured.
                         trace!("received `Start` event when in {} state.", self.state);
                         Effects::new()
                     }
-                    State::Done => {
+                    State::Done(_) => {
                         // Illegal states for syncing start.
                         error!("should not have received `Start` event when in `Done` state.",);
                         Effects::new()
                     }
-                    State::SyncingDescendants {
-                        ref latest_block, ..
-                    } => {
+                    State::SyncingDescendants { latest_block, .. } => {
                         let next_block_height = latest_block.height() + 1;
                         info!(?next_block_height, "start synchronization");
                         self.metrics.reset_start_time();
                         fetch_block_at_height(effect_builder, init_peer, next_block_height)
                     }
-                    State::SyncingTrustedHash { trusted_hash, .. } => {
+                    State::SyncingTrustedHash {
+                        trusted_hash,
+                        highest_block_header,
+                        ..
+                    } => {
                         trace!(?trusted_hash, "start synchronization");
-                        // Start synchronization.
-                        self.metrics.reset_start_time();
-                        fetch_block_by_hash(effect_builder, init_peer, trusted_hash)
+                        match highest_block_header.as_ref() {
+                            Some(hdr) if hdr.era_id().0 > 0 => effect_builder
+                                .get_switch_block_at_era_id_from_storage(EraId(hdr.era_id().0 - 1))
+                                .event(|block| {
+                                    Event::GotInitialValidators(
+                                        init_peer,
+                                        block
+                                            .expect("should have switch block for the latest era")
```

Comment:
(So this doesn't support emergency restarts yet — this will be `None` for emergency restarts?)

Reply:
Doesn't support emergency restart as originally designed - but our current thinking is that we might roll back to the last switch block, which would mean that such restarts are supported.